### ISSUE TITLE
False positive "Using directive is unnecessary" in partial classes

### ISSUE THREAD

**Issue by ulrichb** [dotnet/roslyn#18348]
**Version Used**: 
VS 2017 (15.0.26228.10) / Roslyn 2.0.0

**Steps to Reproduce**:

Partial class file 1:
```C#
using RoslynCodeAnalysisTests.Attributes;

namespace RoslynCodeAnalysisTests
{
    public partial class PartialClass
    {
        partial void PartialMethod([MyAttr] string a)
        {
        }
    }
}
```

Partial class file 2:
```C#
using RoslynCodeAnalysisTests.Attributes;

namespace RoslynCodeAnalysisTests
{
    [MyAttr]
    public partial class PartialClass
    {
        partial void PartialMethod(string a);
    }
}
```

=> Roslyn shows "Using directive is unnecessary" in file 1 for `using RoslynCodeAnalysisTests.Attributes` which is wrong.

![image](https://cloud.githubusercontent.com/assets/388796/24554311/59f8c318-162d-11e7-9ff8-1dfdc1755065.png)

**Expected Behavior**:
Do not show "Using directive is unnecessary". It makes my compilation failing when doing this.

BTW: When executing the corresponding quick-fix it doesn't change anything.

**Comment by Pilchie** [comment]
@CyrusNajmabadi - can you take a look and see if this is a presentation issue, or if it's a problem with the diagnostic reported by the compiler?

**Comment by miloush** [comment]
Well, if it helps

```C#
using RoslynCodeAnalysisTests.Attributes;

namespace RoslynCodeAnalysisTests
{
    public partial class PartialClass
    {
        partial void PartialMethod([MyAttr] string a)
        {
        }
    }
}

namespace RoslynCodeAnalysisTests
{
    [MyAttr]
    public partial class PartialClass
    {
        partial void PartialMethod(string a);
    }
}
```

Invoking _Move to PartialClass.cs_ on the second class removes the using and so breaks the first one...

**Comment by miloush** [comment]
..and the 'add using' quick fix is not available on the parameters's attribute to fix it.

**Comment by CyrusNajmabadi** [comment]
Currently not reproing this:

![image](https://cloud.githubusercontent.com/assets/4564579/24571451/17e8fce2-1626-11e7-86b3-869d23e94c12.png)

![image](https://cloud.githubusercontent.com/assets/4564579/24571452/1c833128-1626-11e7-8295-74411be06a48.png)

It's possible we had a bug binding attributes of partials that was fixed.

**Comment by CyrusNajmabadi** [comment]
I am able to repro the situation where the using is removed.

**Comment by ulrichb** [comment]
> It's possible we had a bug binding attributes of partials that was fixed.

Do you mean that it was fixed after Roslyn 2.0.0?

**Comment by CyrusNajmabadi** [comment]
I have no idea.  I'm still investigating :)

**Comment by ulrichb** [comment]
OK. Thx.

**Comment by CyrusNajmabadi** [comment]
@gafter There is some sort of issue happening over in the compiler.  As part of the "move type" feature, we first pull out the type to be moved, and then we go and remove imports if they are in the new file, and are considered unused by the compiler for the current file.

To determine if the 'using' is unused, we just call into the SemanticModel to get diagnostics, and we look for the CS8019 diagnostics.  In this case, for the updated file:

```c#
using RoslynCodeAnalysisTests.Attributes;

namespace RoslynCodeAnalysisTests
{
    public partial class PartialClass
    {
        partial void PartialMethod([MyAttr] string a)
        {
        }
    }
}
```

the compiler *does* report that the using is unnecessary.  It clearly is necessary as it is used in [MyAttr].

**Comment by CyrusNajmabadi** [comment]
Continuing to investigate.  I want to see why the compiler does'nt mark the using as used for MyAttr.

**Comment by CyrusNajmabadi** [comment]
Ok, i think i see the problem.  @gafter in this scenario, the IDE gets the SemanticModel for a document and then asks GetDiagnostics on it.  That works by calling the following:

```
 	Microsoft.CodeAnalysis.CSharp.MethodCompiler.CompileMethodBodies(compilation, moduleBeingBuiltOpt, generateDebugInfo, hasDeclarationErrors, diagnostics, filterOpt, cancellationToken)	C#
 	Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsForMethodBodiesInTree(tree, span, cancellationToken)	C#
 	Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetDiagnosticsForSyntaxTree(stage, syntaxTree, filterSpanWithinTree, includeEarlierStages, cancellationToken)	C#
 	Microsoft.CodeAnalysis.CSharp.SyntaxTreeSemanticModel.GetDiagnostics(span, cancellationToken)	C#
>	Microsoft.CodeAnalysis.CSharp.RemoveUnnecessaryImports.AbstractCSharpRemoveUnnecessaryImportsService.GetUnnecessaryImports(model, root, predicate, cancellationToken)	C#

```

However, nothing about this "forces" the compiler to resolve the attributes on the parameters of the methods it is compiling.  Because the compiler never bothers to resolve the [MyAttr] attribute on the parameter, the "using" is never marked as used, and as such the compiler reports a diagnostic saying it is unused.

I can imagine a couple of fixes here:

1. When compiling method bodies, we call GetAttributes on all the method parameters to ensure they're resolved properly.
2. Call ForceComplete on the parameters to do the same.

However, i'm not sure what the right thing is to do.  I'm going to assign this to compiler to take a look. 

Thanks!

**Comment by CyrusNajmabadi** [comment]
@gafter Please assign accordingly.  Thanks!

**Comment by jcouv** [comment]
@gafter I didn't want to leave this un-assigned in 15.3. Please advise on priority/assignment.

**Comment by gafter** [comment]
I believe this is not a regression, but a longstanding bug. As such moving it to 15.6. Please let me know if that is wrong and we can reprioritize.

**Comment by wgutierrezr** [comment]
I have this issue in _ViewImports.cshtml in a ASP.Net Core 2.0 project. After the VS-2017 upgrade to 15.4.1

Regards.

**Comment by CyrusNajmabadi** [comment]
HI @wgutierrezr .  

1. Can you provide repro case for us to look at?  
2. Neal only fixed this 23 days ago.  It's unlikely to have made it into 15.4 (it would likely be 15.5 at the earliest).

**Comment by wgutierrezr** [comment]
Hi @CyrusNajmabadi 
My company policy doesn't allow me to send you any type of information. I'm working on a project ASP.Net Core 2.0 MVC with VS2017 Pro, and I had to copy _ViewImports.cshtml for every View folder I have in my project. After upgrade VS2017, when I open any _ViewImport file it shows the code in gray and the message when you have the mouse pointer on any line of code.

It doesn't block me to run the project.

**Comment by DaClan008** [comment]
I think I might have a similar issue on a new Blazor (Wasm) app .Net 5.  I have not made any new updates to the project.  I will test to see if the same happens on a .Net 3.1 version and report back.

My current Visual Studio version is 16.8.2.  I am not a professional developer, so if someone can point me in the right direction to fix this, it will be much appreciated.

I believe I also don't have intellisense at the moment.

Below is a screenshot to show my current running version together with the "warnings" I'm getting.  Please note that it seems the warnings escalate over time (i.e. starts off with 11, then 21 etc. 

![image](https://user-images.githubusercontent.com/8425436/100421389-f6dc7900-3090-11eb-9eeb-d154fd6869cd.png)

**Comment by CyrusNajmabadi** [comment]
Those `usings` look unnecessary as it does not appear that anything needs them (at least as far as i can tell from your screenshot) @DaClan008

**Comment by DaClan008** [comment]
@CyrusNajmabadi, that's interesting.  I have not gone through them all to see if they are actually used.  This is how the project starts up.  

As explained, I have not done a lot of Blazor development (my background is more javascript (node) and some wpf work in the past).  But I thought that the _imports.razor file is used together with "Pages" and "Shared" and if an item is referred to here, you do not have to insert a using statement in the Page / shared item?

FetchData.razor page for instance use the Http.GetFromJsonAsync function which is part of the System.Net.Http namespace, yet there is a message for line 1 of the _import.razor too (in fact there is a message for every item on there).

Is my thinking behind _imports.razor wrong?  Should I remove all using statements, and rather include it in the "page / shared component)?

**Comment by CyrusNajmabadi** [comment]
I could def be wrong here. @NTaylorMullen ?

**Comment by DaClan008** [comment]
This issue is currently resolved by running a repair on visual studio.  (not sure what caused the issue though).

**Comment by NTaylorMullen** [comment]
> This issue is currently resolved by running a repair on visual studio. (not sure what caused the issue though).

Unfortunately this will most likely come back up for you. I'm unsure why a repair fixed; however, the usings are indeed "unnecessary". That being said we're working on fixing this in Razor's new experimental editor.

![image](https://user-images.githubusercontent.com/2008729/87990285-8b5a1d80-ca98-11ea-914f-949fa326d9e3.png)
